lesson3/Main.java

```diff
@@ -0,0 +1,77 @@
+package lesson3;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Random;
+
+public class Main {
+    public static void main(String[] args) {
+        List<Student> students = generateStudents();
+        List<Student> filteredStudents = students.stream()
+                .filter(student -> student.address().city().equals("New York"))
+                .filter(student -> student.age() > 15)
+                .toList();
+
+        List<StudentGradeRecord> studentGradeRecord = filteredStudents.stream()
+                .flatMap(student ->
```

Comment:
There is no need to introduce temporary collection
```suggestion
        List<Student> filteredStudents = students.stream()
                .filter(student -> student.address().city().equals("New York"))
                .filter(student -> student.age() > 15)
                .flatMap(student ->
```